### ISSUE TITLE
Raytrace rendering bug with large objects

### ISSUE THREAD

**Issue by peteihis** [ArtOfIllusion/ArtOfIllusion#193]
There is a bug in ray tracing that is triggered by the size of the object. After some limit size rays seem to return only the color of the ambient light from the surface ignoring the portion of light that should come form lights. 

The triggering size depends on surface type (triangles, or primitive) and there are other factors that affect the result like camera orientation, lighting direction number of lights, distance form the camera to the surface.... It seems that all light produce their own patterns of missing light but sometimes when a pixel rejects light, it rejects it from all lights. Sometimes just one. Another interesting finding is that on a sphere the pattern is mirrored in Y-direction.

![Sphere](https://user-images.githubusercontent.com/22843957/82940303-4b495380-9f9d-11ea-94c6-ff4801b0bd7b.png)

Sphere, R = 80

![Cylinder](https://user-images.githubusercontent.com/22843957/82940315-4f757100-9f9d-11ea-9701-39b6a9832bc2.png)

Cylider, R = 80, H = 80 

![Pyramid_5](https://user-images.githubusercontent.com/22843957/82940329-53a18e80-9f9d-11ea-8ab9-0db3562fc097.png)

Pentagon based pyramid, Edge = 2000

A cube would have to be aroud 10000 units in dimensions before the bug appears. On the pyramid camera distance seems to have a pretty random effect on how bad it gets... Just one Alt+scroll-tick may have the defect almost disappear or ruin the entire image. The case above is somewhat intermediate.

On the outside this seems to be a part of the same family that has produced colorless pixels in the past but there is the difference that the pixels do have a color which allows further processing. The colorless pixels had the tendency of infecting the neighboring pixels as well if for instance filters were applied. Anti aliasing may mask the problem in milder cases but not when the image is noticeably damaged.

Now the process of collecting information from the surface for a ray to return is a pretty complex operation and for the time being I can't begin to imagine where it goes wrong. Some directions, where to dig, would be welcome.

**Comment by Sailsman63** [comment]
Huh...

Findings:

1. This is not *random* noise. for a given object/light/camera position setup, the exact pattern is repeatable. I've even saved a file, and re-started to make sure.
2. Primitive objects do not seem to be affected by surface accuracy settings, or by switching to low-memory mode.
3. The patterns seem to have a very strong directional component. For Spheres, it seems to be related to Camera View direction. For Cylinders, it seems to be related to the rotation of the cylinder itself.

**Comment by Sailsman63** [comment]
I don't really know where to start looking here, but a few things in general if you choose to try to chase this rabbit: (I've done a deep debug in the raytracer once before)

- You're almost certainly going to need to single-step through code at some point, or at least put in some breakpoints. You will either need to use an IDE, or JDB
- The multi-threading that the raytracer does can confuse things a lot. You want to force the use of a single thread to filter out a lot of noise. Usually, on Windows, I just set processor affinity to a single core through Task Manager before I started the render process.

**Comment by peteihis** [comment]
And it gets better.....

Above it looks like the bug only ignored light, but I just tested with the depth map script and that that too produces black pixels. This can only mean that some rays just happily travelled through the object, missing it existence entirely. The object was 5000 × 8000 × 5000 units with flat outer sides.

Also I had another test case, where I tried to raytrace a depth map of an imported stl. I left it running on Friday afternoon and it was still "running" this morning, but the process had died some time during the weekend.... So far I'm not sure what happens there. Simpler objects may get rendered/depthmapped, but sometimes with "holes" as above. I tried importing a smaller portion of the model but that did not help. I tried removing triangles that are smaller than some size limit, but that did not help. (Reduced triangel count drastically though). It still may be just the mere size of the objects, that is the problem... Need to test it in smaller scale.

**Comment by peteihis** [comment]
I was thinking that I'd need to isolate a single pixel that fails an see what happens in every step of the way, when it is traced. Did not get through checking the entire process yet, though.

**Comment by Sailsman63** [comment]
>just tested with the depth map script and that that too produces black pixels. This can only mean that some rays just happily travelled through the object

If this is true: (Sounds right so far) That narrows things down a bit. 

Most probable at this point:

Either some miss-sorting in the octree, or rounding errors in the `checkIntersection()` implementations.

**Comment by peteihis** [comment]
Some more edges:

I'm not sure how to repeat that black pixels in the depth map thing... I'll try it again with the same computer and create the object the same way (imported STL). Somehow it seems that there may be up to three different bugs here or maybe the same one appearing in three different ways:

- The pixels missing the light component
- Rendering falling into infinity loop at start (Process hangs before all the threads are even up. CPU may stay running at 28% or something like that...)
- The failed pixels in the depth map.

Had a test case that got rendered with quality issues, but pulling the camera back just one click, threw the next attempt to infinity loop. On the very edge before the loop case stranger things may happen:

- With a two object scene, one object could appear in the image and the other one not.
- One of the objects is a scripted CSG-object. It may happen that the first object of the operation is rendered instead of the final piece. (This time the object that was show in the case before was missing from the image.)
- Sometimes the first Ctrl+Shift+R renders the image but trying it again, without changing anything loops the process. -- As if the first rendering had somehow not been finished properly though it managed to finish the image....

The distance from the camera to the object is a factor in all of the cases and the depth mapper falls into loop at the same distance as the renderer.

A rounding error sounds like a credible suspect. (Thinking of a case where iteration starts to produce the same value (or zic-zac with two values) and there is fixed tolerance value in wrong decades, that is then never met....)

**Comment by Sailsman63** [comment]
>Rendering falling into infinity loop at start (Process hangs before all the threads are even up. CPU may stay running at 28% or something like that...)

If you can get it to do that again, please grab a Thread Dump or two. (VisualVM is usually the easiest - https://visualvm.github.io/download.html as it's no longer included with the JDK)

The dumps can be saved as files and attached, rather than Copy/Paste into the thread.

**Comment by Sailsman63** [comment]
>With a two object scene, one object could appear in the image and the other one not.

When you raytrace, or in the interactive view? Interactive renderers do sometimes do weird stuff if you get too close to the camera...

The others there seem bizarre.

- Any log output?
- Is this a release version, or a self-compile of one of our development versions?
    + If self compiled, make sure that you run `ant clean` before rebuilding, especially if you have switched branches. I've had some weird stuff happen when I forget to do that...

**Comment by peteihis** [comment]
> When you raytrace, or in the interactive view?

This is all about raytracing.

Nothing out of the ordinary in the logs.

Now at home, the latest nighty version. At work the current head. It was compiled in a copy of the (branched) git folder, so no wrong version build files could be in the build. _(When the development folders need cleaning I usually just delete the build-folder. I don't like to do builds in the repository that communicates with Github, because I usually end up with all sort of side tracks and experimental versions that I want to keep away from the "final" stuff... Even so I have trashed an 'uncountable number' of branches gone wrong over time.)_

> Thread Dump

Will try.

**Comment by peteihis** [comment]
... and there actually seem to be three separate problems.

Just to test I changed the `TOL = 1e-12` that `RTTriangle.checkIntersection` uses to `1e-6`. That fixed the pixels that were missing the light, but pulling the camera back still loops the ray tracing and black pixels still appear when they do. The black pixels are actually happening on a primitive Cube (which does not necessarily exclude other objects)

The `TOL` is basically used with [0, 1] normalized values but this looks like a very potential place for a rounding error issue. Got to do the same check-ups on the rest of the RT objects.

**Comment by peteihis** [comment]
I tested the process hang with the depth mapper. 

What happens in those cases is that `RayTracer.traceRay()` is called and it never returns. Based on that it probably is that all threads do get started and the lucky ones get finished too. Only the threads, that run into a problem, stay looping somewhere in the intersection detection and the process just keeps waiting.

There were also some black pixels on the partial image I got from that experiment, which seems to hint, that there were some close calls but it managed to return maximum distance instead of looping.

**Comment by Sailsman63** [comment]
What line number is shown for the `Raytracer.traceRay()`call? Does it change if you take several thread dumps?

**Comment by peteihis** [comment]
> Have you managed to find a consistent way to force this hang to happen?

I've got a case that has it happen. It is really just about the camera distance and the viewing angle that need to exceed some limit value. With some (temporary) mods I managed to put all three different error modes into one picture.

![Defects 320 × 320](https://user-images.githubusercontent.com/22843957/83480551-625bda00-a4a4-11ea-934e-a4c516bece28.png) ![Deph_map 320 × 320](https://user-images.githubusercontent.com/22843957/83480554-638d0700-a4a4-11ea-9236-f28c3cc60931.png)

The color picture: On the closest side to camera some pixels missing the light component. Rays "through" the object on the left corner and inside. (Not really through, but rather just not detecting the surface.)

The depth map picture: Transparent pixels with background color would show up without color but allow rendering to continue. The black ones are of the type that loop the rendering thread. For these pictures I forced exiting the loop by a counter.

The loop that the program does not get out of is the `while` loop in the second `traceRay()`. (I did not thread dump I found it otherwise.) The conditions, that should result exiting the loop seem pretty solid to me but somehow a failure to detect an intersection confuses it. Triangles that reach closer to camera than the surface that actually is shown somehow seem to affect the result. 

Need to go deeper to how the intersection is detected. It seems to be a numerical problem of some sort but the strange part is, that the rays are just direction vectors starting from a point in space and I'd assume those to be immune to distance pretty much up to `Double.MAX_VALUE`.

**Comment by peastman** [comment]
Just as a guess, I suspect this is related to the code for handling self intersections.  When a ray hits a surface, you need to spawn rays from the hit point to all the lights so you can identify which ones are illuminated.  But due to round-off error, sometimes the first surface one of those rays "hits" will actually be the surface you're starting it from.  In principle the ray origin is exactly on the surface, but due to finite precision it will actually be a tiny distance away from the surface, and that could cause it to hit the surface after traveling a finite distance.  So you need to identify those spurious hits and exclude them, which turns out to be surprisingly difficult to do robustly.

**Comment by Sailsman63** [comment]
That would definitely make sense of the pixels that are only showing the ambient lighting.

We're talking about larger than typical objects, so many point coordinates will be a large absolute distance from the scene origin, which will cause the effective precision to be lower than it would for a more typical test scene.

A similar error could cause a ray to skip through entirely, which would account for the background showing through.

**Comment by peastman** [comment]
> One way would be to check if the surface, that the the ray found is actually the same one it just left...

If the surface it's starting from is an RTTriangle that should work.  A ray can only intersect a flat triangle once, so any intersection it finds is guaranteed to be the one that's supposed to be at the ray origin.

For any other type of object you can't immediately exclude it because a ray really can intersect the object at more than one place.  A robust (but possibly inefficient) solution would be to then trace a ray in the opposite direction.  Whichever is the closest self intersection it finds, in either the positive or negative direction, is the one we want to ignore.

**Comment by peteihis** [comment]
I guess in most cases you'd be dealing with triangles, so a little inefficiency with the others would not necessarily be a disaster. 

How about checking the direction of the surface normal? If it is the same (or _generally_ the same for curved surfaces) as at the intersection the ray is leaving you assume that the ray is still in intersection?

**Comment by peastman** [comment]
I don't think that would be reliable.  For example, a sphere can be viewed either from the inside or the outside.

**Comment by peteihis** [comment]
> For example, a sphere can be viewed either from the inside or the outside.

Ehhh.... I don't see why not? If the surface normal on a point on the surface of a sphere is the same as it was the last time, you obviously are at the same spot. It does not matter which direction the ray came form, the normal at any point on the surface is unique to that point.

**Comment by peastman** [comment]
Oh, I see what you mean.  You aren't comparing the surface normal to the ray direction.  You're comparing it to the surface normal at the previous intersection point.

That might well work.  At least, I haven't thought of any situations where it would fail.  There are cases where the normals would be exactly opposite each other (opposite faces of a cube), but I don't see any where they'd be parallel.

**Comment by peteihis** [comment]
Some test results:

The short version is that Peter was right about how the mechanism how some pixels end up ignoring the light and that the way to deal with it that was discussed above would work. However, that would only be a way to live with the problem not to solve it.

I created a new test case: Two spheres, one with the default texture and another one with a blue/cyan mirror surface, cyan on the mirror and blue as the diffuse component.

Here's what happens with 3.1.0:

![TestCase_TwoSpheres](https://user-images.githubusercontent.com/22843957/83948160-274f0300-a824-11ea-9650-830d568247ae.png)

What may be a bit unexpected is that, when the radius of the spheres goes from 100 to 1000000 the problem does not get much worse. The spots start to appear at R = 39 and, unlike with flat surfaces, the camera distance has very little effect on the result. It depends almost solely on the object size.

In the other end of the scale, at R = 0,0000001, the spheres seem to disappear into the error margin and the light rays never leave the surface.

Then I made the error trap into `RaytracerRenderer.traceLightRay()`. Here's what happens:

![BS2_1e6](https://user-images.githubusercontent.com/22843957/83948354-6d589680-a825-11ea-93cc-1919df3a683f.png)

These are the million-unit spheres. The error trap would have to be built for every ray that attempts to bounce off an object surface... Lots of them in the code and this does not help in the small end.

Then I refined an older idea a bit further: The root cause to the problem is the fixed `TOL = 1e-12` setting everywhere. This works fine as long as dimensions are within a decade or two from 1.0. As a matter of fact for spheres and cylinders even the 2nd decade is too much.

So instead of TOL I used `tol = r2*TOL` (r2 = r*r in the code). Then things started to happen:

![ModC](https://user-images.githubusercontent.com/22843957/83948543-a9402b80-a826-11ea-91f3-b38a3cc22ffe.png)

So basically the relative tolerance would add flexibility and we could, in theory go "from atomic to astronomical", quite effortlessly. Though, when I tried anything bigger than R = 1e6, something else started to go wrong and the image might or might not have been rendered.... At least something was left running in the memory and the only way out was to kill the Java process.

I also tested things like `tol = Math.sqrt(r2*3)*TOL` but that only helped a couple of decades up in scale. The numerical error grows in powers as numbers are multiplied. What exactly is the needed "power of tolerance" probably needs more study, but that seems to be the correct principle.

**Comment by peteihis** [comment]
![Importacelow](https://user-images.githubusercontent.com/22843957/83949404-405bb200-a82c-11ea-958f-c27bf9a2137f.png)

Just playing with the thought, but as this began with the largeish mm-scale import, the 1e13-size would already be closer to "nuclearic" rather than "atomic" in millimeters. In the astronomy end of the scale, if we consider 200000 ly to be the diameter of the Milky Way it would translate to 1,89e24 mm. A nice target for scale handling. :)

**Comment by Sailsman63** [comment]
What was the code for the "Error Trap" that you tried? (Just out of curiosity)

**Comment by peteihis** [comment]
> What was the code for the "Error Trap" that you tried?

~~~~Java
    ....
    int i, j, matCount = 0;

    RTObject firstObject = workspace.firstObjectHit;  // new
    Vec3 objectTrueNormal = workspace.trueNormal[0];  // new
    Vec3 newTrueNormal = new Vec3();  // new

    do
    {
      RTObject obj[] = node.getObjects();
      for (i = obj.length-1; i >= 0; i--)
      {
        SurfaceIntersection intersection = r.findIntersection(obj[i]);
        intersection.trueNormal(newTrueNormal);

        // This is the part that does the work
        if (obj[i] == firstObject && objectTrueNormal.distance2(newTrueNormal) < 0.002) // new
          continue;  // new

        if (intersection != SurfaceIntersection.NO_INTERSECTION)
        ....
~~~~

So checking, if the new found object and it's surface normal are the same ones the first ray portion hit.

**Comment by Sailsman63** [comment]
Did that check make a difference with the triangle message variant to this issue?

**Comment by peteihis** [comment]
Yes. It removed the missing light pixels but no effect on the other problems.

I Got already a little bit further: I set size dependent `tol` values for some of the RTObjects and that does the same thing. _(Did not find the exact same camera angle as before.)_

![c1](https://user-images.githubusercontent.com/22843957/83956094-fb06a700-a862-11ea-91e0-8b330c3407e3.png) ![c2](https://user-images.githubusercontent.com/22843957/83956096-ffcb5b00-a862-11ea-885b-f52595bd53e5.png)

Helps with the missing light but not, when the rays miss the objects entirely. It started to look like it had something to with the bounding boxes of `OctreeNodes`, so I tested what happens if I turn the test object around its y-axis ... Well:

![Cubes_pic](https://user-images.githubusercontent.com/22843957/83955677-2aff7b80-a85e-11ea-8d33-1996ff43f750.png)
![Cubes_map](https://user-images.githubusercontent.com/22843957/83955723-ef18e600-a85e-11ea-932b-de5038357691.png)

I guess that'd be a safe assumption about the bounding boxes.. :D

What you see is all in one scripted object that I turned 45° in the scene. The black pixels are again result of the counter interrupted while loop. By the normal process intersection distance is forced to 0.0 for `NO_INTERSECTION` but on those it is `Double.POSITIVE_INFINITY`.

When I had those as two separate objects, it could happen that the renderer only drew one of them... Nodes having a box fight...

Now the `OctreeNode` again uses `Raytracer.TOL` for some purposes.... I don't know yet what does what there, but this may be a trickier case as the bug seems to be depending on the distance between the ray origin and the world origin and possibly the ray length too.

**Comment by Sailsman63** [comment]
> It removed the missing light pixels but did not affect the other problems.

> I got already a little bit further: I set size dependent tol values for some of the RTObjects and that does the same thing. 


That confirms that we have two different, though similar, bugs

I'd be glad to see a PR with the scaling-tolerance solution for all the standard `RTObject` types to poke at. It looks like it whould be much cheaper than the conditional trap, and should be more robust as well.

**Comment by Sailsman63** [comment]
>The black pixels are again result of the counter interrupted while loop.

Where, exactly, did you put this counter? I was going to ask this before, but forgot about it. That is definitely a different issue.

**Comment by peteihis** [comment]
> Where, exactly, did you put this counter?

I thought I had mentioned, but is in the second `Raytracer.traceRay()` -- The one that returns `OctreeNode`. For this purpose the while loop is forced exit after 500 cycles.

Anyway here's a little debugging package, for anyone who cares: Both of the `traceRay`s and a not very stable version of the `DepthMapRenderer`-script that can visualize the problems that the `traceRays` find.

[RayTracer debugging.zip](https://github.com/ArtOfIllusion/ArtOfIllusion/files/4742760/RayTracer.debugging.zip)

This is what they do:

![FailureModes](https://user-images.githubusercontent.com/22843957/83977411-07453f80-a909-11ea-913e-d361513ac0ba.png)

Explanation:
- Cyan pixels: Ray is never traced because no "**_first node_**" is found.
- Blue pixel: The 'intersect.distance' is never given a value. The while loop is exited normally by the `(first == null)` and `(node == null)` condition.
- Red pixels: Appears in the infinity loop case. There is no **_first object_** for the ray but a **_next node_** will always appear.

And I got one bit forward again. I'll skip the eventful but extremely boring history of how I ended up there, but the reason for the transparent pixels on objects was, that `OctreeNode.findFirstNode()` started to suffer from numerical errors after a pretty sharp limit distance. When it did not find the node, the ray was never traced. Adding a bit of tolerance cleared that problem. 

![findFirstNode](https://user-images.githubusercontent.com/22843957/83978058-80df2c80-a90d-11ea-8c3b-ad7c53c4c608.png)

I'll still have to figure out how the error behaves when the scale changes, but let's say that the root cause is identified and the principle for the solution is known.

The while loop never exiting seems to be caused by a similar but inverted problem in the `OctreeNode.findNextNode()`, so that when the error happens, the method always generates a new node without realizing that there is nothing left....  I have not started to study that yet.

Then there is at least one more infinity loop to find. I don't know yet where exactly to look but it appears when you are far away enough from the world origin and start to trace lights. On top of that there are quite a few `while` and `do` loops in the code. Not all of them are potential hazards but any one that handles object dimensions or coordinates might be.

**Comment by peteihis** [comment]
> I'd be glad to see a PR with the scaling-tolerance solution for all the standard RTObject types

Yes, I thought that it could be PR:ed as a case of it's own. At the moment the other issues are a little bit on the way so all the RTOs can not be tested to equal extremities to really see how the error behaves.

I'll be at it. :)

**Comment by Sailsman63** [comment]
>I'll be at it. :)

No rush... after all, this is a hobby project for all of us.

**Comment by peteihis** [comment]
>  .... this is a hobby

"A hobby isn't a real hobby, if it does not take all you have..."

But it seems, that I got the major `OctreeNode` scaling limits somehow sorted out and now "the other infinity loop" did not occur any more with the test case I had for it. I was able to test a (scripted object) primitive cube from 1e-14 to 1e128 edge lengths _(did not test how high I could go though)_ with no pixel issues and no infinity loops....

With more complex objects the realistic upper size limit seems to be about 1e6 but the restrictions don't come from the raytracer renderer. For example a primitive sphere can be rendered with R = 1e24 but when you open it for editing the next time, the content of the number fields is no longer 1e24... (it was about 9e-15, with a long string of decimals) though the size of the sphere had not changed.  As a scripted object the sphere R >= 1e7 loops something else and the UI freezes with the script editor open. I don't know what happens when there are more objects in the scene either.  

Anyway, that part looks pretty good, I'll put the RTObjects through some kind of a stress test next.

**Comment by peteihis** [comment]
It seems that I have hit a little snag somewhere... I made a test case with a single RTTriangle`, flat on xy-plane and that seems to have a different idea from the rest of what works or not. I'm not yet sure what to do with it, but I'll try to avoid special cases if possible.

One root of evil with `OctreeNode` that the `minx ... maxz` are float's and the position vector that those are compared to comes in `doubles`. At around `1e7` dimensions (or, as it seems, `1e-7` as well) the relative error does no longer comply with the `1e-12` tolerance. The `contains` method that does the comparing is called from two other methods and unfortunately as adding a tolerance value to the comparison solves one problem it the induces another problem in the other case. 

In practical terms, getting rid of rays that miss the object indices an infinity loop in the other method. (What I said about the infinity loop earlier was not exactly like that. It's more like that the `findNextNode` just kept returning this node, when it should have returned "nothing found".

But anyways something positive. In most cases it seems possible to reach `1e32` dimensions which I would call "astronomical enough". That is a bit over 100 times larger than the diameter of the Observable Universe in millimeters. These experiments also reveal some (maybe a bit surprising too) short comings of the UI. For example auto zoomig from 1e32 directly to 1e-6 pushes the camera view into an unrecoverable error mode... I'm not sure if the problem is with the view, the camera or the animation engine...

To be continued...

**Comment by peastman** [comment]
It's supposed to be handling the float/double conversion correctly by always being conservative.  See `findMinimum()` and `findMaximum()` in OctreeNode.  Given a BoundingBox, which has a range specified in double precision, it calls `Math.nextAfter()` to find the nearest single precision value just outside the double precision range.  So it errs on the side of including a little more inside the node than it needs to.

**Comment by peteihis** [comment]
It probably does handle the conversion correctly. Difficulties begin, when the object OctreeNode is very thin and the ray too long to calculate the intersection point accurately enough enough to end up inside the node.

Actually this phenomenon seems to be more complex than I would have thought, see the .mov in the .zip. The width and height of the triangle are 6E8 units, it is placed in the world origin and the camera FOV is 45°.

[trngl_009.zip](https://github.com/ArtOfIllusion/ArtOfIllusion/files/4787990/trngl_009.zip)

**Comment by peteihis** [comment]
Some kind of a mile stone achieved! I think this update belongs here rather than the PR-thread.

All the before test cases work now and my current version RTRenderer can now handle rather a wide range of dimensions. For this picture I used 1 unit = 1 meter scale, the cube represents a dust particle 0.1 mm in size and the Earth in the back is in the actual correct dimensions, 12713600 meters tall and 12756200 meters wide. _(And I have no idea which way the Milky Way should have been tilted...)_

![Earth and dust](https://user-images.githubusercontent.com/22843957/87128878-136e4500-c299-11ea-8387-edb5f134fd09.jpg)

Of course this kind of a scene does not work very well with objects that need triangulation. To get the near by objects smoothed the large ones would be subdivided into very small chips.... Maybe another project some time.

One strange thing is still left: With primitives there is no problem, but when there is a triangulated object it only works close to world origin. The object close to camera is about 0.001 units in size (= 1 mm), the sphere about the size of Earth. On the right side image I moved everything 10 meters right an up. The bounding box starts to cut slices off the space somewhere after 3 meters. This happens in some form with or without the large object in the back, though it has some effect on the subdivision of the box.

![At_C](https://user-images.githubusercontent.com/22843957/87129219-a7d8a780-c299-11ea-9ef5-257c00a93494.png) ![Off_C](https://user-images.githubusercontent.com/22843957/87129280-b8891d80-c299-11ea-8dd5-46f7f816fd46.png)

Some kind of a hint of what is happening is that this happens per direction. If I move the objects in x-direction only, I get only upright walls that cut the space... 

And another finding:

![Sea_001a](https://user-images.githubusercontent.com/22843957/87130952-5a116e80-c29c-11ea-9e88-7e70aeecb24d.jpg)

Some black pixels in the sea. They really are black, not background color or ambient color as in the above cases. In the texture there is a ViewAngle module that is set to absolute value. The setting did not make uch difference though.

**Comment by peastman** [comment]
Wow.  That's seriously impressive!

**Comment by peteihis** [comment]
> Wow. That's seriously impressive!

Thank you! :)

**Comment by peteihis** [comment]
That sea scene case is somewhat mysterious: This seems to be specific to Cylinders. I tested with a Cube in similar proportions and that was working right.

On those rays that painted a black pixel the `trueNorm` points _into_ the object surface, not away from it. That is because the `RTCylinder` in those cases thinks that the ray hit the bottom of the cylinder instead of the top. (!) Somehow the texture is a factor too  though it should not. This version of the texture is otherwise the same as above but without transparency, so it shows the problem worse.

![clsea](https://user-images.githubusercontent.com/22843957/87252665-94863180-c47d-11ea-8cac-503610a8f23e.png) ![cbesea](https://user-images.githubusercontent.com/22843957/87252667-98b24f00-c47d-11ea-9fbe-8c408fd87035.png)

Left: Cylinder sea, right Cube sea.

Somehow the thickness (height) or the cylinder and the y-position are factors as well... Actually hitting just those numbers, that cause that effect, on purpose does not seem easy at all.... But it still should not happen.
